File: services/worker-py/src/jobforge_worker/handlers/http_request.py
"""HTTP Request Connector - Python implementation."""

import re
import time
from typing import Any
from urllib.parse import urlparse

import httpx
from pydantic import BaseModel, Field, field_validator


class HttpRequestPayload(BaseModel):
    """HTTP request job payload."""

    url: str
    method: str = "GET"
    headers: dict[str, str] = Field(default_factory=dict)
    body: str | dict[str, Any] | None = None
    timeout_ms: int = Field(default=30_000, ge=1, le=300_000)
    allowlist: list[str] = Field(default_factory=list)
    redact_headers: list[str] = Field(
        default_factory=lambda: ["authorization", "cookie", "set-cookie"]
    )

    @field_validator("method")
    @classmethod
    def validate_method(cls, v: str) -> str:
        """Validate HTTP method."""
        allowed = ["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"]
        if v.upper() not in allowed:
            raise ValueError(f"Method must be one of {allowed}")
        return v.upper()

    @field_validator("url")
    @classmethod
    def validate_url(cls, v: str) -> str:
        """Basic URL validation."""
        if not v.startswith(("http://", "https://")):
            raise ValueError("URL must start with http:// or https://")
        return v


class HttpRequestResult(BaseModel):
    """HTTP request result."""

    status: int
    duration_ms: int
    response_headers: dict[str, str]
    response_body_preview: str
    success: bool


BLOCKED_HOSTS = {
    "localhost",
    "127.0.0.1",
    "0.0.0.0",
    "169.254.169.254",  # AWS metadata
    "metadata.google.internal",  # GCP metadata
}

PRIVATE_IP_PATTERNS = [
    re.compile(r"^10\."),
    re.compile(r"^172\.(1[6-9]|2[0-9]|3[0-1])\."),
    re.compile(r"^192\.168\."),
    re.compile(r"^127\."),
    re.compile(r"^0\."),
]


def validate_url_ssrf(url: str, allowlist: list[str]) -> None:
    """
    SSRF protection: validate URL against allowlist and block private IPs.

    Args:
        url: Target URL
        allowlist: List of allowed host patterns

    Raises:
        ValueError: If URL is blocked or not in allowlist
    """
    parsed = urlparse(url)
    hostname = parsed.hostname or parsed.netloc

    # Check blocked hosts
    if hostname.lower() in BLOCKED_HOSTS:
        raise ValueError(f"Blocked host: {hostname}")

    # Check private IP ranges
    for pattern in PRIVATE_IP_PATTERNS:
        if pattern.match(hostname):
            raise ValueError(f"Private IP address not allowed: {hostname}")

    # Check allowlist if provided
    if allowlist:
        allowed = False
        for pattern in allowlist:
            if "*" in pattern:
                regex = re.compile("^" + pattern.replace("*", ".*") + "$")
                if regex.match(hostname):
                    allowed = True
                    break
            elif hostname == pattern or hostname.endswith(f".{pattern}"):
                allowed = True
                break

        if not allowed:
            raise ValueError(f"Host not in allowlist: {hostname}")


def http_request_handler(payload: dict[str, Any], context: dict[str, Any]) -> dict[str, Any]:
    """
    HTTP request handler.

    Args:
        payload: Job payload
        context: Job context

    Returns:
        Request result

    Raises:
        ValueError: If validation fails
        httpx.HTTPError: If request fails
    """
    validated = HttpRequestPayload.model_validate(payload)

    # SSRF protection
    validate_url_ssrf(validated.url, validated.allowlist)

    start_time = time.time()

    with httpx.Client(timeout=validated.timeout_ms / 1000.0) as client:
        # Prepare request body
        body_data = None
        if validated.body and validated.method not in ["GET", "HEAD"]:
            body_data = validated.body

        # Make request
        response = client.request(
            method=validated.method,
            url=validated.url,
            headers=validated.headers,
            json=body_data if isinstance(validated.body, dict) else None,
            content=body_data if isinstance(validated.body, str) else None,
        )

    duration_ms = int((time.time() - start_time) * 1000)

    # Redact sensitive headers
    response_headers = {}
    for key, value in response.headers.items():
        if key.lower() not in validated.redact_headers:
            response_headers[key] = value

    # Read response body with size limit
    max_body_size = 1_000_000  # 1MB
    body_text = response.text
    if len(body_text) > max_body_size:
        response_body_preview = body_text[:max_body_size] + "... (truncated)"
    else:
        response_body_preview = body_text

    result = HttpRequestResult(
        status=response.status_code,
        duration_ms=duration_ms,
        response_headers=response_headers,
        response_body_preview=response_body_preview,
        success=response.is_success,
    )

    return result.model_dump()

File: services/worker-ts/src/handlers/http-request.ts
/**
 * HTTP Request Connector
 * Executes HTTP requests with SSRF protection including DNS-rebinding prevention
 */

import type { JobContext } from '@jobforge/shared'
import { z } from 'zod'
import { promises as dns } from 'node:dns'
import { isIP } from 'node:net'
import { getAllowlistMatcher } from './allowlist-matcher'
import { collectResponseHeaders } from './response-headers'
import { readBodyPreview } from './response-preview'

const HttpRequestPayloadSchema = z.object({
  url: z.string().url(),
  method: z.enum(['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'HEAD']).default('GET'),
  headers: z.record(z.string()).optional(),
  body: z.union([z.string(), z.record(z.unknown())]).optional(),
  timeout_ms: z.number().int().positive().max(300_000).default(30_000),
  allowlist: z.array(z.string()).optional(),
  redact_headers: z.array(z.string()).default(['authorization', 'cookie', 'set-cookie']),
  response_headers_allowlist: z.array(z.string()).optional(),
})

export type HttpRequestPayload = z.infer<typeof HttpRequestPayloadSchema>

export interface HttpRequestResult {
  status: number
  duration_ms: number
  response_headers: Record<string, string>
  response_body_preview: string
  success: boolean
}

const BLOCKED_HOSTNAMES = [
  'localhost',
  '127.0.0.1',
  '0.0.0.0',
  '169.254.169.254', // AWS metadata
  'metadata.google.internal', // GCP metadata
  'instance-data',
]

/**
 * Validates whether an IP address belongs to a private, loopback, or link-local range.
 */
export function isPrivateIp(ip: string): boolean {
  const version = isIP(ip)
  if (!version) return true // Invalid IP considered unsafe

  if (version === 4) {
    const parts = ip.split('.').map((p) => parseInt(p, 10))
    if (parts.length !== 4 || parts.some((p) => isNaN(p) || p < 0 || p > 255)) {
      return true
    }

    const [a, b] = parts

    // 0.0.0.0/8
    if (a === 0) return true
    // 10.0.0.0/8
    if (a === 10) return true
    // 127.0.0.0/8 (Loopback)
    if (a === 127) return true
    // 169.254.0.0/16 (Link-local / Cloud metadata)
    if (a === 169 && b === 254) return true
    // 172.16.0.0/12 (172.16.0.0 - 172.31.255.255)
    if (a === 172 && b >= 16 && b <= 31) return true
    // 192.168.0.0/16
    if (a === 192 && b === 168) return true
    // 100.64.0.0/10 (Carrier-grade NAT)
    if (a === 100 && b >= 64 && b <= 127) return true
    // 224.0.0.0/4 (Multicast) or 240.0.0.0/4 (Reserved)
    if (a >= 224) return true

    return false
  }

  if (version === 6) {
    const lower = ip.toLowerCase()
    // Loopback ::1
    if (lower === '::1' || lower === '0:0:0:0:0:0:0:1') return true
    // Unspecified ::
    if (lower === '::' || lower === '0:0:0:0:0:0:0:0') return true
    // IPv4-mapped IPv6 (::ffff:x.x.x.x)
    if (lower.startsWith('::ffff:')) {
      const ipv4Part = lower.replace('::ffff:', '')
      if (isIP(ipv4Part) === 4) {
        return isPrivateIp(ipv4Part)
      }
      return true
    }
    // Unique local address fc00::/7 (fc00... or fd00...)
    if (lower.startsWith('fc') || lower.startsWith('fd')) return true
    // Link-local fe80::/10
    if (
      lower.startsWith('fe8') ||
      lower.startsWith('fe9') ||
      lower.startsWith('fea') ||
      lower.startsWith('feb')
    ) {
      return true
    }

    return false
  }

  return true
}

/**
 * SSRF protection: validate URL against allowlist and resolve DNS to prevent DNS-rebinding
 */
export async function validateUrl(url: string, allowlist?: string[]): Promise<void> {
  const parsed = new URL(url)
  const hostname = parsed.hostname.toLowerCase()

  // Only allow http and https protocols
  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
    throw new Error(`Unsupported protocol: ${parsed.protocol}`)
  }

  // Check static blocked hostnames
  if (BLOCKED_HOSTNAMES.includes(hostname)) {
    throw new Error(`Blocked host: ${hostname}`)
  }

  // Check allowlist if provided
  const matcher = getAllowlistMatcher(allowlist)
  if (matcher && !matcher(hostname)) {
    throw new Error(`Host not in allowlist: ${hostname}`)
  }

  // If hostname is already a direct IP address
  if (isIP(hostname)) {
    if (isPrivateIp(hostname)) {
      throw new Error(`Private IP address not allowed: ${hostname}`)
    }
    return
  }

  // Resolve hostname via DNS to prevent DNS rebinding attacks
  try {
    const addresses = await dns.lookup(hostname, { all: true })
    if (!addresses || addresses.length === 0) {
      throw new Error(`DNS resolution returned no records for host: ${hostname}`)
    }

    for (const record of addresses) {
      if (isPrivateIp(record.address)) {
        throw new Error(
          `Host ${hostname} resolves to prohibited private IP address: ${record.address}`
        )
      }
    }
  } catch (err) {
    if (err instanceof Error && err.message.includes('prohibited private IP address')) {
      throw err
    }
    throw new Error(
      `DNS resolution failed for host ${hostname}: ${err instanceof Error ? err.message : String(err)}`
    )
  }
}

/**
 * HTTP Request Handler
 */
export async function httpRequestHandler(
  payload: unknown,
  _context: JobContext
): Promise<HttpRequestResult> {
  const validated = HttpRequestPayloadSchema.parse(payload)

  // SSRF protection with DNS resolution check
  await validateUrl(validated.url, validated.allowlist)

  const startTime = Date.now()

  try {
    const response = await fetch(validated.url, {
      method: validated.method,
      headers: validated.headers,
      body:
        validated.body && validated.method !== 'GET' && validated.method !== 'HEAD'
          ? typeof validated.body === 'string'
            ? validated.body
            : JSON.stringify(validated.body)
          : undefined,
      signal: AbortSignal.timeout(validated.timeout_ms),
    })

    const _duration_ms = Date.now() - startTime

    // Redact sensitive headers
    const response_headers = collectResponseHeaders(response, {
      redactHeaders: validated.redact_headers,
      allowlist: validated.response_headers_allowlist,
    })

    // Read response body with size limit
    const MAX_BODY_SIZE = 1_000_000 // 1MB
    const bodyPreview = await readBodyPreview(response, MAX_BODY_SIZE)

    return {
      status: response.status,
      duration_ms: _duration_ms,
      response_headers,
      response_body_preview: bodyPreview.bodyPreview,
      success: response.ok,
    }
  } catch (error) {
    throw new Error(
      `HTTP request failed: ${error instanceof Error ? error.message : String(error)}`
    )
  }
}
